SConstruct

```diff
@@ -62,154 +81,98 @@ if GetOption("base_config") == "e2_xml":
+        # Convert entry to plain string so there are no differences
+        target = str(target_raw)
+        # between plain commandline targets and "option" targets
+        multienv[target] = env.Clone(
+            BUILD_LABEL=target, BUILD_DIR=os.path.abspath(target)
         )
-    )
-    env["LINKFLAGS"] = cproject.get_link_flags(e2_target)
-    env["LINKCOM"] = "$CXX $CPPFLAGS -o $TARGET $SOURCES $LINKFLAGS"
-    
-    
-    env["LIBPATH"] = [str(i) for i in cproject.get_link_libs_order(e2_target)]
-    # log.debug(env["LIBPATH"])
+
+        if not multienv[target].SConscript(
+            os.path.join("site_scons", "e2_xml.scons"),
+            exports={
+                "env": multienv[target],
+                "log_parent": log,
+                "cproject": cproject,
+                "e2_build_targets": e2_build_targets,
+                "dbt_build_targets": dbt_build_targets,
+            },
+        ):
+            success = False
+            Return("success")
 
 # GENERIC ENVIRONMENT STUFF AND ACTIONS
 # In this area below the "if" block, we keep environment changes and actions that are
 # going to need doing regardless of the particular "mode" (eg: e2_xml).
 
-# Helper functions borrowed from FBT for wrapping lengthy commandlines into temp files.
-# This is primarily useful in windows, but shouldn't hurt anything on other platforms.
-wrap_tempfile(env, "LINKCOM")
-wrap_tempfile(env, "ARCOM")
-
-# Happy little custom builders and their actions.
-env.Append(
-    BUILDERS={
-        "HEXBuilder": Builder(
-            action=Action(
-                '${OBJCOPY} -O ihex "${SOURCE}" "${TARGET}"',
-                "${HEXCOMSTR}",
-            ),
-            suffix=".hex",
-            src_suffix=".elf",
-        ),
-        "BINBuilder": Builder(
-            action=Action(
-                '${OBJCOPY} -O binary -S "${SOURCE}" "${TARGET}"',
-                "${BINCOMSTR}",
-            ),
-            suffix=".bin",
-            src_suffix=".elf",
-        ),
-    }
-)
+# Iterate again, this time over configured environments. This should work regardless of
+# --base_config setting!
+for target, t_env in multienv.items():
+    # Helper functions borrowed from FBT for wrapping lengthy commandlines into temp files.
+    # This is primarily useful in windows, but shouldn't hurt anything on other platforms.
+    wrap_tempfile(t_env, "LINKCOM")
+    wrap_tempfile(t_env, "ARCOM")
+
+    # VariantDir does the magic to ensure output goes to the dbt- whatever
+    # build directory. Careful: THIS CAN BE REALLY FINICKY if paths aren't set right
+    VariantDir(
+        os.path.join(t_env["BUILD_LABEL"], REL_SOURCE_DIR), "#src", duplicate=False
+    )
 
-# Export a `compile_commands.json` for help
-env.Tool('compilation_db')
-env.CompilationDatabase()
+    # Export a `compile_commands.json` for help
+    # t_env["COMPILATIONDB_PATH_FILTER"] = t_env["BUILD_LABEL"]
+    t_env["COMPILATIONDB_USE_ABSPATH"] = True
```

Comment:
Can move this to the default env, probably? I think the `Tool` invocation can be as well but I'm not a SCons expert.

Reply:
It *can* definitely go into the default env, but my hope was to use the default env as a starting point for other functionality like debugging that may not be handling a target in the "build" sense. Since it's build target related, I put it in the t_env (badly named "generic target environment").